### PR TITLE
CAC-11604: Platform Usage (Load) panel for Science Portal

### PR DESCRIPTION
**Pull request by hjeeves**
Changes:
- progress bars have more granularity (one per panel, as 2 panels can now be refreshed separately.)
- page state can be grabbed from the react app in order to set progress bars
- science_portal_session.js javascript contains calls to get platform usage (skaha/session?view=stats)
- new SciencePortalPlatformUsage react component to display values
- changes in App.js for the extra panel

Technical debt changes:
- indentation in App.js was severely off in some areas
- remove commented code in App.js, science_portal_core.js, science_portal.js

Tested manually on a canfar vm

### CODE REVIEW

File: TECH_README.md [hunk 1]
```diff
@@ -12,7 +12,8 @@ react developer tools in chrome: https://chrome.google.com/webstore/detail/react
 host file entry (see Environment Files)
 
 ### Environment files
-.env file needs to be set up in the root of your local copy of the code, with contents similar to:
+.env file needs to be set up in the root of your local copy of the code, with contents similar to
```
Comment (at88mph) [inline TECH_README.md]: It actually just needs to be in the same place as the `docker-compose.yml` file.

Or more appropriately:

> Relative path MUST be resolved from the Compose file’s parent folder.

File: public/dev/js/science_portal.js [hunk 2]
```diff
@@ -78,6 +74,7 @@
       portalCore.setHeaderURLs()
       portalForm.setContentBase(contentBase)
       portalForm.loadSessionTypeMap()
+      portalCore.setPageState("all", "primary", true, '')
```
Comment (at88mph) [inline public/dev/js/science_portal.js]: Mix of quotes here.

File: public/dev/js/science_portal.js [hunk 3]
```diff
@@ -230,7 +255,7 @@
           "listType" : "empty",
           "sessData" : []
         }
-        _reactApp.updateSessionList(sessDataObj)
+        var currentState = _reactApp.updateSessionList(sessDataObj)
```
Comment (at88mph) [inline public/dev/js/science_portal.js]: This variable isn't used anywhere.

File: public/dev/js/science_portal_core.js [hunk 4]
```diff
@@ -150,65 +157,79 @@
...
+      return progressBar
+    }
 
-      if ((typeof alertMsg !== "undefined") && (alertMsg !== "")) {
-        pageState.alert = {
-            "type": barType,
-            "show":  true,
-            "message": alertMsg
+    // Communicate AJAX progress and status using progress bar
+    // function setPageState(barType, isAnimated, formMsg, sessionActionAlertMsg) {
```
Comment (at88mph) [inline public/dev/js/science_portal_core.js]: This can be removed.

File: public/dev/js/science_portal_core.js [hunk 5]
```diff
@@ -150,65 +157,79 @@
...
+          // Used to reset entire page, or to
+          // set all progress bars to red if there's a skaha service error
+          pageState.spForm.alert.show = false
+          pageState.spForm.progressBar = _mkProgressBarData(barType, isAnimated)
+          pageState.spPlatformUsage.alert.show = false
+          pageState.spPlatformUsage.progressBar = _mkProgressBarData(barType, isAnimated)
+          pageState.spSessionList.alert.show = false
+          pageState.spSessionList.progressBar = _mkProgressBarData(barType, isAnimated)
+        } else {
+          // pageState[pageSection].alert = _mkAlertData(barType, )
```
Comment (at88mph) [inline public/dev/js/science_portal_core.js]: This can be removed.

File: public/dev/js/science_portal_core.js [hunk 6]
```diff
@@ -150,65 +157,79 @@
...
+
+          // var show = true
```
Comment (at88mph) [inline public/dev/js/science_portal_core.js]: This can be removed.

File: public/dev/js/science_portal_session.js [hunk 7]
```diff
@@ -259,12 +278,76 @@
...
+                  biggestCount = val
+                }
+              }
+            });
+
+            // This will be used for the max height of the bar chart being displayed
+            // Code is here rather than in the SciencePortalPlatformUsage component
+            // because it's better to do this work once than (potentially)
+            // every time the component is rendered
+            // var chartHeight = ((biggestCount + 10) % 10 ) * 10
```
Comment (at88mph) [inline public/dev/js/science_portal_session.js]: This can be removed.

File: public/dev/js/science_portal_session.js [hunk 8]
```diff
@@ -275,30 +358,31 @@
...
-          false
-        )
-        // withCredentials enables cookies to be sent
-        request.withCredentials = true
-        request.open("GET", serviceURL)
-        request.send(null)
-      })
-    }
+    //
+    // function getSessionListAjax(serviceURL, sessionData) {
```
Comment (at88mph) [inline public/dev/js/science_portal_session.js]: This can be removed.

File: public/dev/js/science_portal_session.js [hunk 9]
```diff
@@ -372,15 +456,41 @@
...
+    }
 
     function pollSessionList(interval) {
         // TODO: consider long-running timeout so panel left in background doesn't use
         // resources forever
         interval = interval || 200
 
         var checkCondition = function (resolve, reject) {
-          getSessionListAjax(_selfPortalSess.sessionServiceURL)
+          // getSessionListAjax(_selfPortalSess.sessionServiceURL)
```
Comment (at88mph) [inline public/dev/js/science_portal_session.js]: This can be removed.

**Review by at88mph — CHANGES_REQUESTED** [review]

File: public/dev/js/science_portal_session.js (hunk 9 above)
Comment (hjeeves) [reply public/dev/js/science_portal_session.js]: good catch. this code got merged into a different more generalized function and I didn't remove the original one!

File: public/dev/js/science_portal_session.js (hunk 8 above)
Comment (hjeeves) [reply public/dev/js/science_portal_session.js]: (as the comment below)

File: public/dev/js/science_portal_core.js (hunk 4 above)
Comment (hjeeves) [reply public/dev/js/science_portal_core.js]: removed the commented function statement & updated the comment.

File: public/dev/js/science_portal_core.js (hunk 5 above)
Comment (hjeeves) [reply public/dev/js/science_portal_core.js]: ok

File: public/dev/js/science_portal_core.js (hunk 6 above)
Comment (hjeeves) [reply public/dev/js/science_portal_core.js]: ok

File: public/dev/js/science_portal_session.js (hunk 7 above)
Comment (hjeeves) [reply public/dev/js/science_portal_session.js]: ok

File: public/dev/js/science_portal.js (hunk 2 above)
Comment (hjeeves) [reply public/dev/js/science_portal.js]: Somehow, always has to be one left!

File: public/dev/js/science_portal.js (hunk 3 above)
Comment (hjeeves) [reply public/dev/js/science_portal.js]: removed.

File: TECH_README.md (hunk 1 above)
Comment (hjeeves) [reply TECH_README.md]: I dug a bit further into what is using this .env file. It's npm, and it's only used to start up the local version. (command in package.json 'sudo react-scripts start', which uses npm under the covers.  The simplest way is to have this .env file at the application root so npm can find it. I added a link to documentation so if anyone wants to have a different configuration (or have different env files for different setups,) it's possible.

**Review by at88mph — APPROVED** [review]

**Pull request merged**